packages/kbn-esql-validation-autocomplete/src/shared/variables.ts

```diff
@@ -35,41 +29,75 @@ function addToVariables(
 ) {
   if (isColumnItem(oldArg) && isColumnItem(newArg)) {
     const newVariable: ESQLVariable = {
-      name: newArg.name,
+      name: newArg.parts.join('.'),
       type: 'double' /* fallback to number */,
       location: newArg.location,
     };
     // Now workout the exact type
     // it can be a rename of another variable as well
-    const oldRef =
-      fields.get(oldArg.name) || variables.get(oldArg.quoted ? oldArg.text : oldArg.name);
+    const oldRef = fields.get(oldArg.parts.join('.')) || variables.get(oldArg.parts.join('.'));
     if (oldRef) {
-      addToVariableOccurrencies(variables, newVariable);
+      addToVariableOccurrences(variables, newVariable);
       newVariable.type = Array.isArray(oldRef) ? oldRef[0].type : oldRef.type;
     }
   }
 }
 
-function getAssignRightHandSideType(item: ESQLAstItem, fields: Map<string, ESQLRealField>) {
-  if (Array.isArray(item)) {
-    const firstArg = item[0];
-    if (Array.isArray(firstArg) || !firstArg) {
-      return;
+/**
+ * Determines the type of the expression
+ *
+ * TODO - this function needs a lot of work. For example, it needs to find the best-matching function signature
+ * which it isn't currently doing.
+ * @param root
+ * @returns
+ */
+function getExpressionType(
+  root: ESQLAstItem,
+  fields: Map<string, ESQLRealField>,
+  variables: Map<string, ESQLVariable[]>
+): string {
+  const fallback = 'double';
+
+  if (Array.isArray(root) || !root) {
+    return fallback;
+  }
+  if (root.type === 'literal') {
+    return root.literalType;
+  }
+  if (root.type === 'inlineCast') {
+    if (root.castType === 'int') {
+      return 'integer';
     }
-    if (firstArg.type === 'literal') {
-      return firstArg.literalType;
+    if (root.castType === 'bool') {
+      return 'boolean';
     }
```

Comment:
Will be nice to automate keeping this in sync in https://github.com/elastic/kibana/issues/186409